### ISSUE TITLE
Failed to raise an exception: 990087232 on compilation (aarch64)

### ISSUE THREAD

**Issue by zethexx** [omarroth/invidious#953]
I am getting:
```
Failed to raise an exception: 990087232
[0xaaaadf3bbe68] ???
[0xaaaadeb907d4] __crystal_raise +40
[0xaaaadeb91c18] ???
[0xaaaadeb9b1d4] ???
[0xaaaadeb9627c] ???
[0xaaaadeb959ac] ???
[0xaaaadee94394] ???
[0xaaaadeba5614] ???
[0xaaaadeb93910] main +60
[0xffff8f542d24] __libc_start_main +228
[0xaaaadeb8f244] ???
```
when trying to build `invidious.cr` with `$ crystal build src/invidious.cr --release` as the invidious user (pwd is ~/invidious)

I successfully got crystal compiler to build with aarch64 and tested it. (crystal -v) and used it to compile shards. I copied both binaries to a directory in my `${PATH}` (`/usr/local/bin`).

However after some time compiling (one core pinned to 100% so it's frozen compiling) it crashes with the above "Failed to raise an exception" error. I don't know why it is happening.

**Comment by tleydxdy** [comment]
if you mean the compiler is crashing; yes, it does that some times, remove cache and build again. if you mean the binary itself, see https://github.com/omarroth/invidious/issues/917 .

**Comment by zethexx** [comment]
The compiler is the one crashing. I tried a second time and I got the same results:
```
Failed to raise an exception: 601914944
[0xaaaae670be68] ???
[0xaaaae5ee07d4] __crystal_raise +40
[0xaaaae5ee1c18] ???
[0xaaaae5eeb1d4] ???
[0xaaaae5ee627c] ???
[0xaaaae5ee59ac] ???
[0xaaaae61e4394] ???
[0xaaaae5ef5614] ???
[0xaaaae5ee3910] main +60
[0xffffb5f4cd24] __libc_start_main +228
[0xaaaae5edf244] ???
```
I will try removing cache and report back.

**Comment by zethexx** [comment]
I got same results after removing `~/.cache/crystal`:
```
Failed to raise an exception: -405094144
[0xaaaab016ee68] ???
[0xaaaaaf9437d4] __crystal_raise +40
[0xaaaaaf944c18] ???
[0xaaaaaf94e1d4] ???
[0xaaaaaf94927c] ???
[0xaaaaaf9489ac] ???
[0xaaaaafc47394] ???
[0xaaaaaf958614] ???
[0xaaaaaf946910] main +60
[0xffff7f78fd24] __libc_start_main +228
[0xaaaaaf942244] ???
```

**Comment by zethexx** [comment]
I figured out how to get past that part. I needed to add llvm-config for llvm-8 to my `$PATH`. But now I am getting error of trying to cc lsquic:
```
/usr/bin/ld: /home/invidious/invidious/lib/lsquic/src/lsquic/ext/liblsquic.a(lsquic_conn.c.o): Relocations in generic ELF (EM: 62)
/usr/bin/ld: /home/invidious/invidious/lib/lsquic/src/lsquic/ext/liblsquic.a(lsquic_conn.c.o): Relocations in generic ELF (EM: 62)
/usr/bin/ld: /home/invidious/invidious/lib/lsquic/src/lsquic/ext/liblsquic.a: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/home/invidious/invidious/invidious'  -rdynamic  /home/invidious/invidious/lib/lsquic/src/lsquic/ext/liblsquic.a -lz -lyaml -lxml2 -lsqlite3 -lz /home/invidious/invidious/lib/lsquic/src/lsquic/ext/libssl.a `command -v pkg-config > /dev/null && pkg-config --libs --silence-errors libssl || printf %s '-lssl -lcrypto'` /home/invidious/invidious/lib/lsquic/src/lsquic/ext/liblsquic.a /home/invidious/invidious/lib/lsquic/src/lsquic/ext/libcrypto.a `command -v pkg-config > /dev/null && pkg-config --libs --silence-errors libcrypto || printf %s '-lcrypto'` -lpcre -lm -lgc -lpthread /home/rock64/crystal-bootstrap/crystal-0.31.1/src/ext/libcrystal.a -levent -lrt -ldl -L/usr/lib -L/usr/local/lib`
```

**Comment by zethexx** [comment]
I'm going to close this issue here because it's probably just something missing from my system. I've heard someone else that got invidious to compile (except failing in cc for lsquic issues that I will discuss later) on a raspberry pi. I've also got it to compile (except failing in cc for lsquic issues) by cross-compiling on my amd64 system. Probably something missing in my aarch64 system.
---
The real problem here with running Invidious on aarch64 and other ARM architectures is the lsquic dependency. The lsquic shard (which pulls https://github.com/omarroth/lsquic) contains already compiled binaries of libcrypto, liblsquick and libssl which are not for aarch64/ARM, only arm64. This explains the ELF format errors I got. I will try investigate for a solution in the near future.
